packages/daemon/DESIGN.md

```diff
@@ -0,0 +1,215 @@
+
+- The program can use the Powers object to request an ephemeral value
+  from the user.
+  The Powers will not retain the resulting formula.
+
+  ```js
+  const file = E(powers).request('a file');
+  ```
+
+- The program can use the Powers object to request a durable value
+  and give it a pet name.
+  The Powers will memoize the value and retain the resulting formula.
+  Endo will honor subsequent requests for the same pet name from this Powers
+  object by reconstructing the value from its formula if necessary.
+
+  ```js
+  const file = E(powers).request('a file, 'fileName');
+  ```
+
+- The user can respond to pending requests by streaming requests
+  from their Inbox.
+  The user can resolve the request with the pet name of a formula they hold,
+  giving the guest the corresponding value.
+  The user can reject the request with a message.
+  The user can observe whether the request is still pending, which will be
+  useful for a graphical user interface.
+
+  ```js
+  for await (const request of iterateRef(E(inbox).requests())) {
+    request.resolve('userPetName');
+  }
+  ```
+
+- The user can create a new Powers object by calling the `makePowers`
+  method of their Inbox.
+  In a user interface, requests will be attributed to their pet name for the
+  Powers object.
+
+Every formula has a string identifier.
+If the behavior of the formula cannot be inferred from the identifier,
+Endo stores a corresponding JSON description in a key value store (tentatively
+just files) indexed on the identifier.
+The description instructs Endo in how to create the corresponding value.
+Formulas with dependencies list the identifiers of those dependencies in a
+`slots` property then refer elsewhere to those formulas by the index of the
+slot.
+This makes formulas transparent to a garbage collector.
+
+- A blob formula denotes a content address hash for bytes that Endo has stored.
+  Endo provides a Blob reference for that formula with methods for reading the
+  text, bytes, or byte stream from the blob.
+
+  Formula Identifier:
+
+  ```
+  readableSha512/$SHA512
+  ```
+
+  Endo stores the blob in the user's home directory, tentatively in a file with
+  the same name as the formula identifier.
+
+- A worker formula denotes the unique identifier for an Endo worker.
+  Endo provides a Worker reference for the process.
+  Workers can execute confined and unconfined programs and have a `terminate`
+  method.
+
+  Formula Identifier:
+
+  ```
+  workerUuid/$UUID
+  ```
+
+  Endo stores the worker's last known PID and a combined stdout and stderr log
+  in a directory with the same name.
+
+- For an evaluation formula, Endo provides the completion value of a JavaScript
+  program to be run in the identified worker, confined to an Endo Compartment,
+  endowed with the values for the identified formulas, bound to the given
+  names.
+  Endo Compartments are endowed with `E` and `Far` for sending and receiving
+  messages.
```

Comment:
Aha! Now I appreciate why you want to make these global rather than imports.

Reply:
These do not need to be global for confined programs, but they do need to be endowed for `eval` formulas since dynamic import would be unwieldy. I should note that eval scripts are very useful for ad hoc orchestration at the command line or a REPL, but are not intended to represent normal usage by end users.